docs/architecture/ADR-002-universal-share-encoding.md

```diff
@@ -0,0 +1,144 @@
+# ADR 009: Universal Share Encoding
+<!-- disable markdownlint MD010 because Go code snippet uses tabs -->
+<!-- markdownlint-disable MD010 -->
+
+## Terminology
+
+- **nid** (8 bytes): namespace id
+- **reserved** (1 byte): is the location of the first transaction, ISR, or evidence in this share if there is one and `0` if there isn't one
+- **message length** (varint 1 to 10 bytes): is the length of the entire message in bytes
+- **compact share**: a type of share that can accomodate multiple units. Currently, compact shares are used for transactions, ISRs, and evidence to efficiently pack this information into as few shares as possible.
+- **sparse share**: a type of share that can accomodate zero or one unit. Currently, sparse shares are used for messages.
+
+## Context
+
+The current compact share format is:
+
+- First share of reserved namespace: <br>`nid (8 bytes) | reserved (1 byte) | data length (varint 1 to 10 bytes) | data`
+- Contiguous share in reserved namespace:<br>`nid (8 bytes) | reserved (1 byte) | data`
+
+The current sparse share format is:
+
+- First share of message:<br>`nid (8 bytes) | message length (varint 1 to 10 bytes) | data`
+- Contiguous share in message:<br>`nid (8 bytes) | data`
+
+The current share format poses multiple challenges:
+
+1. Clients must have two share parsing implementations (one for compact shares and one for sparse shares).
+1. It is difficult to make changes to the share format in a backwards compatible way because clients can't determine which version of the share format an individual share conforms to.
+1. It is not possible for a client that samples a random share to determine if the share is the first share of a message or a contiguous share in the message.
+
+## Proposal
+
+Introduce a universal share encoding that applies to both compact and sparse shares:
+
+- First share of namespace for compact shares or message for sprase shares:<br>`nid (8 bytes) | info (1 byte) | data length (varint 1 to 10 bytes) | data`
+- Contiguous share in namespace for compact shares or message for sparse shares:<br>`nid (8 bytes) | info (1 byte) | data`
+
+Shares in the reserved namespace have the added constraint: the first byte of `data` is a reserved byte so the format is:<br>`nid (8 bytes) | info (1 byte) | data length (varint 1 to 10 bytes) | reserved (1 byte) | data`
+
+Where **info** (1 byte) is a byte with the following structure:
+
+- the first 7 bits are reserved for the version information in big endian form (initially, this will be `0000000` for version 0);
+- the last bit is a **message start indicator**, that is `1` if the share is at the start of a message or `0` if it is a contiguous share within a message.
```

Comment:
[naming] The message start indicator will be included in the `InfoReservedByte` that is used for compact shares so it seems appropriate to drop the "message" from this name.

[question] when should the start indicator be `1` for a compact share? Two options:
1. For the first share in the namespace (i.e. the first transaction share in the transaction namespace)
2. For any share in the namespace that contains the start of a unit. (i.e. for the transaction namespace, in any share where the reserved byte is non `0`)

Tagging @evan-forbes if you have insight on the above. The current implementation [ref](https://github.com/celestiaorg/celestia-app/pull/691/files#diff-3da7e413fef3c90c01286110ba9bdfc3d61bbb360355bb32adede4e7b3ebfe5bR27) handles **option 1** but it can be modified if option 2 is preferred.

Reply:
The important thing in an implementation would be to treat everything as that struct. So we could not name it different, and then treat everything as a "Message". I don't really have a firm opinion yet, dropping the message makes sense too.

> [question] when should the start indicator be 1 for a compact share? Two options:

yes definitely 1 and not 2. 

I don't think this is occurring, but tbc on wording for other reviewers
> For the first share in the namespace

is only for the current implementation of info byte for _compact_ shares, the implementation of universal shares will obvi not only be for the first share in a namespace

Reply:
well shoot, maybe we need to change the wording in the adr to make that clearer? the entire point of universal shares is that we treat the splitting of each namespace identically for the first step, so saying

For the first share in the namespace

could be confusing.

Reply:
> The fact that there is only a single message in reserved namespaces is purely an application detail 

Agreed. 

Since compact shares contain a reserved byte, we need some way of distinguishing between sparse shares and compact shares. In other words, since compact shares are an application detail, parsing them requires separate logic from sparse shares. Although the API to `ParseShare` can be used for both compact and sparse shares, the API still needs to be able to differentiate the type of share being parsed which I assume will be based on the namespace id.

Note: we can enforce an additional requirement that there is only one compact share (i.e. the first) per reserved namespace with the `start indicator=1`.

But the terminology used in this ADR is confusing with respect to "For the first share in the namespace", will try to improve.

Reply:
>the API still needs to be able to differentiate the type of share being parsed which I assume will be based on the namespace id

Why is that?

Reply:
I think to parse the reserved byte that is only encoded into compact shares

Reply:
That reserved byte should be part of the raw share data so there shouldn't be a need to parse it to parse a share

Reply:
Agreed parsing the reserved byte is only necessary after parsing the universal prefix. 

I think it's possible to improve the terminology here but given other readers are familiar with "message" as describing the contents of a share (even if the share contains non-message data, i.e. transactions), I think it's safe to resolve my thread on [naming]

Reply:
Unresolving b/c I think the naming is still confusing. What are reviewers thoughts on renaming the message start indicator to payload start indicator or data start indicator? @evan-forbes @musalbas 

The rename is motivated by an attempt at conveying that the start indicator is applicable to shares without message data (i.e. compact shares).